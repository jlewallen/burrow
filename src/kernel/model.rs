use anyhow::Result;
use nanoid::nanoid;
use once_cell::sync::Lazy;
use serde::{Deserialize, Serialize};
use std::{
    cell::RefCell,
    collections::HashMap,
    fmt::{Debug, Display},
    ops::{Deref, DerefMut, Index},
    rc::{Rc, Weak},
};
use thiserror::Error;
use tracing::{debug, span, trace, Level};

use super::{infra::*, Scope};

pub static WORLD_KEY: Lazy<EntityKey> = Lazy::new(|| EntityKey("world".to_string()));

pub static NAME_PROPERTY: &str = "name";

pub static DESC_PROPERTY: &str = "desc";

pub static GID_PROPERTY: &str = "gid";

#[derive(Clone)]
pub struct EntityPtr {
    entity: Rc<RefCell<Entity>>,
    lazy: LazyLoadedEntity,
}

impl EntityPtr {
    pub fn new_blank() -> Self {
        let brand_new = Rc::new(RefCell::new(Entity::default()));
        let lazy = LazyLoadedEntity::new_from_raw(&brand_new);

        Self {
            entity: brand_new,
            lazy,
        }
    }

    pub fn downgrade(&self) -> Weak<RefCell<Entity>> {
        Rc::downgrade(&self.entity)
    }

    pub fn key(&self) -> EntityKey {
        self.lazy.key.clone()
    }
}

impl From<Rc<RefCell<Entity>>> for EntityPtr {
    fn from(ep: Rc<RefCell<Entity>>) -> Self {
        let lazy = LazyLoadedEntity::new_from_raw(&ep);

        Self {
            entity: Rc::clone(&ep),
            lazy,
        }
    }
}

impl From<Entity> for EntityPtr {
    fn from(entity: Entity) -> Self {
        Rc::new(RefCell::new(entity)).into()
    }
}

impl Debug for EntityPtr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("EntityPtr")
            .field("key", &self.lazy.key)
            .field("name", &self.lazy.name)
            .finish()
    }
}

// This seems cleaner than implementing borrow/borrow_mut ourselves and things
// were gnarly when I tried implementing Borrow<T> myself.
impl Deref for EntityPtr {
    type Target = RefCell<Entity>;

    fn deref(&self) -> &Self::Target {
        self.entity.as_ref()
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Serialize, Deserialize, Debug)]
pub struct EntityGID(i64);

impl EntityGID {
    pub fn new(i: i64) -> EntityGID {
        EntityGID(i)
    }

    pub fn gid_to_string(&self) -> String {
        format!("{}", self.0)
    }
}

impl Display for EntityGID {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl From<EntityGID> for i64 {
    fn from(gid: EntityGID) -> Self {
        gid.0
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
pub struct EntityKey(String);

impl EntityKey {
    pub fn new(s: &str) -> EntityKey {
        EntityKey(s.to_string())
    }

    pub fn key_to_string(&self) -> &str {
        &self.0
    }
}

impl Debug for EntityKey {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("`{}`", self.0))
    }
}

impl Default for EntityKey {
    fn default() -> Self {
        Self(nanoid!())
    }
}

impl Display for EntityKey {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

pub trait DomainEvent: Debug {}

#[derive(Debug)]
pub enum DomainOutcome {
    Ok(Vec<Box<dyn DomainEvent>>),
    Nope,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Item {
    Named(String),
    Route(String),
    GID(EntityGID),
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct EntityRef {
    #[serde(rename = "py/object")]
    py_object: String,
    #[serde(rename = "py/ref")]
    py_ref: String,
    pub key: EntityKey,
    #[serde(rename = "klass")]
    class: String,
    name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct Identity {
    #[serde(rename = "py/object")]
    py_object: String,
    private: String,
    public: String,
    signature: Option<String>, // TODO Why does this happen in the model?
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Kind {
    #[serde(rename = "py/object")]
    py_object: String,
    identity: Identity,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct EntityClass {
    #[serde(rename = "py/type")]
    py_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AclRule {
    #[serde(rename = "py/object")]
    py_object: String,
    keys: Vec<String>,
    perm: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct Acls {
    #[serde(rename = "py/object")]
    py_object: String,
    rules: Vec<AclRule>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Version {
    #[serde(rename = "py/object")]
    py_object: String,
    i: u32,
}

impl Default for Version {
    fn default() -> Self {
        Self {
            py_object: Default::default(),
            i: 1,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Property {
    #[serde(rename = "py/object")]
    py_object: String,
    acls: Acls,
    value: serde_json::Value,
}

impl Property {
    pub fn new(value: serde_json::Value) -> Self {
        Self {
            py_object: "".to_string(),
            acls: Default::default(),
            value,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Props {
    #[serde(rename = "py/object")]
    py_object: String,
    map: HashMap<String, Property>,
}

impl Default for Props {
    fn default() -> Self {
        Self {
            py_object: "model.properties.Common".to_string(), // #python-class
            map: Default::default(),
        }
    }
}

impl Props {
    fn property_named(&self, name: &str) -> Option<&Property> {
        if self.map.contains_key(name) {
            return Some(self.map.index(name));
        }
        None
    }

    fn string_property(&self, name: &str) -> Option<String> {
        if let Some(property) = self.property_named(name) {
            match &property.value {
                serde_json::Value::String(v) => Some(v.to_string()),
                _ => None,
            }
        } else {
            None
        }
    }

    fn i64_property(&self, name: &str) -> Option<i64> {
        if let Some(property) = self.property_named(name) {
            match &property.value {
                serde_json::Value::Number(v) => v.as_i64(),
                _ => None,
            }
        } else {
            None
        }
    }

    fn set_property(&mut self, name: &str, value: serde_json::Value) {
        self.map.insert(name.to_string(), Property::new(value));
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct Entity {
    #[serde(rename = "py/object")]
    py_object: String,
    pub key: EntityKey,
    version: Version,
    parent: Option<LazyLoadedEntity>,
    creator: Option<LazyLoadedEntity>,
    identity: Identity,
    #[serde(rename = "klass")]
    class: EntityClass,
    acls: Acls,
    props: Props,
    scopes: HashMap<String, serde_json::Value>,

    #[serde(skip)] // Very private
    infra: Option<Weak<dyn Infrastructure>>,
}

impl Display for Entity {
    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
        fmt.debug_struct(&self.class.py_type)
            .field("key", &self.key)
            .field("name", &self.name())
            .field("gid", &self.gid())
            .finish()
    }
}

impl Needs<std::rc::Rc<dyn Infrastructure>> for Entity {
    fn supply(&mut self, infra: &std::rc::Rc<dyn Infrastructure>) -> Result<()> {
        self.infra = Some(Rc::downgrade(infra));
        self.parent = infra.ensure_optional_entity(&self.parent)?;
        self.creator = infra.ensure_optional_entity(&self.creator)?;
        Ok(())
    }
}

impl Entity {
    pub fn name(&self) -> Option<String> {
        self.props.string_property(NAME_PROPERTY)
    }

    pub fn set_name(&mut self, value: &str) -> Result<()> {
        let value: serde_json::Value = value.into();
        self.props.set_property(NAME_PROPERTY, value);

        Ok(())
    }

    pub fn desc(&self) -> Option<String> {
        self.props.string_property(DESC_PROPERTY)
    }

    pub fn gid(&self) -> Option<EntityGID> {
        if let Some(id) = self.props.i64_property(GID_PROPERTY) {
            Some(EntityGID(id))
        } else {
            None
        }
    }

    pub fn has_scope<T: Scope>(&self) -> bool {
        self.scopes.contains_key(<T as Scope>::scope_key())
    }

    pub fn scope_mut<T: Scope>(&mut self) -> Result<OpenScopeMut<T>, DomainError> {
        let scope = self.load_scope::<T>()?;
        Ok(OpenScopeMut::new(self, scope))
    }

    pub fn scope<T: Scope>(&self) -> Result<OpenScope<T>, DomainError> {
        let scope = self.load_scope::<T>()?;
        Ok(OpenScope::new(scope))
    }

    pub fn maybe_scope<T: Scope>(&self) -> Result<Option<OpenScope<T>>, DomainError> {
        if !self.has_scope::<T>() {
            return Ok(None);
        }
        let scope = self.load_scope::<T>()?;
        Ok(Some(OpenScope::new(scope)))
    }

    fn load_scope<T: Scope>(&self) -> Result<Box<T>, DomainError> {
        let scope_key = <T as Scope>::scope_key();

        let _load_scope_span = span!(
            Level::DEBUG,
            "scope",
            key = self.key.key_to_string(),
            scope = scope_key
        )
        .entered();

        if !self.scopes.contains_key(scope_key) {
            return Ok(Box::new(Default::default()));
        }

        // The call to serde_json::from_value requires owned data and we have a
        // reference to somebody else's. Presumuably so that we don't couple the
        // lifetime of the returned object to the lifetime of the data being
        // referenced? What's the right solution here?
        // Should the 'un-parsed' Scope also owned the parsed data?
        let data = &self.scopes[scope_key];
        let owned_value = data.clone();
        let mut scope: Box<T> = serde_json::from_value(owned_value)?;

        let _prepare_span = span!(Level::DEBUG, "prepare").entered();

        if let Some(infra) = &self.infra {
            if let Some(infra) = infra.upgrade() {
                scope.supply(&infra)?;
            } else {
                return Err(DomainError::ExpiredInfrastructure);
            }
        } else {
            return Err(DomainError::NoInfrastructure);
        }

        Ok(scope)
    }

    fn replace_scope<T: Scope>(&mut self, scope: &T) -> Result<()> {
        let scope_key = <T as Scope>::scope_key();

        let _span = span!(
            Level::DEBUG,
            "scope",
            key = self.key.key_to_string(),
            scope = scope_key
        )
        .entered();

        let value = scope.serialize()?;

        debug!("scope-replace");

        self.scopes.insert(scope_key.to_string(), value);

        Ok(())
    }
}

pub struct OpenScope<T: Scope> {
    target: Box<T>,
}

impl<T: Scope> OpenScope<T> {
    pub fn new(target: Box<T>) -> Self {
        trace!("scope-open {:?}", target);

        Self { target }
    }
}

impl<T: Scope> Deref for OpenScope<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        self.target.as_ref()
    }
}

pub struct OpenScopeMut<'me, T: Scope> {
    owner: &'me mut Entity,
    target: Box<T>,
}

impl<'me, T: Scope> OpenScopeMut<'me, T> {
    pub fn new(owner: &'me mut Entity, target: Box<T>) -> Self {
        trace!("scope-open {:?}", target);

        Self { owner, target }
    }

    pub fn save(&mut self) -> Result<()> {
        self.owner.replace_scope(self.target.as_ref())?;

        Ok(())
    }
}

impl<'me, T: Scope> Drop for OpenScopeMut<'me, T> {
    fn drop(&mut self) {
        // TODO Check for unsaved changes to this scope and possibly warn the
        // user, this would require them to intentionally discard  any unsaved
        // changes. Not being able to bubble an error up makes doing anything
        // elaborate in here a bad idea.
        trace!("scope-dropped {:?}", self.target);
    }
}

impl<'me, T: Scope> Deref for OpenScopeMut<'me, T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        self.target.as_ref()
    }
}

impl<'me, T: Scope> DerefMut for OpenScopeMut<'me, T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        self.target.as_mut()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct LazyLoadedEntity {
    #[serde(rename = "py/object")]
    py_object: String,
    #[serde(rename = "py/ref")]
    py_ref: String,
    pub key: EntityKey,
    #[serde(rename = "klass")]
    class: String,
    name: String,

    #[serde(skip)]
    entity: Option<Weak<RefCell<Entity>>>,
}

impl LazyLoadedEntity {
    pub fn new_with_entity(entity: EntityPtr) -> Self {
        Self::new_from_raw(&entity.entity)
    }

    fn new_from_raw(entity: &Rc<RefCell<Entity>>) -> Self {
        let shared_entity = entity.borrow();
        Self {
            py_object: "model.entity.EntityRef".to_string(), // #python-class
            py_ref: "model.entity.Entity".to_string(),       // #python-class
            key: shared_entity.key.clone(),
            class: shared_entity.class.py_type.clone(),
            name: shared_entity.name().unwrap_or_default(),
            entity: Some(Rc::downgrade(entity)),
        }
    }

    pub fn has_entity(&self) -> bool {
        self.entity.is_some()
    }

    pub fn into_entity(&self) -> Result<EntityPtr, DomainError> {
        match &self.entity {
            Some(e) => match e.upgrade() {
                Some(e) => Ok(e.into()),
                None => Err(DomainError::DanglingEntity),
            },
            None => Err(DomainError::DanglingEntity),
        }
    }
}

impl From<EntityPtr> for LazyLoadedEntity {
    fn from(entity: EntityPtr) -> Self {
        entity.lazy
    }
}

#[derive(Error, Debug)]
pub enum DomainError {
    #[error("no such scope '{:?}' on entity '{:?}'", .0, .1)]
    NoSuchScope(EntityKey, String),
    #[error("parse failed")]
    ParseFailed(#[source] serde_json::Error),
    #[error("dangling entity")]
    DanglingEntity,
    #[error("anyhow error")]
    Anyhow(#[source] anyhow::Error),
    #[error("no infrastructure")]
    NoInfrastructure,
    #[error("expired infrastructure")]
    ExpiredInfrastructure,
    #[error("session closed")]
    SessionClosed,
    #[error("container required")]
    ContainerRequired,
}

impl From<serde_json::Error> for DomainError {
    fn from(source: serde_json::Error) -> Self {
        DomainError::ParseFailed(source)
    }
}

impl From<anyhow::Error> for DomainError {
    fn from(source: anyhow::Error) -> Self {
        DomainError::Anyhow(source)
    }
}

#[derive(Error, Debug)]
pub enum EvaluationError {
    #[error("parse failed")]
    ParseFailed,
}

impl From<nom::Err<nom::error::Error<&str>>> for EvaluationError {
    fn from(source: nom::Err<nom::error::Error<&str>>) -> EvaluationError {
        match source {
            nom::Err::Incomplete(_) => EvaluationError::ParseFailed,
            nom::Err::Error(_e) => EvaluationError::ParseFailed,
            nom::Err::Failure(_e) => EvaluationError::ParseFailed,
        }
    }
}
