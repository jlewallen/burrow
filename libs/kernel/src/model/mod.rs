use anyhow::Result;
use std::{
    cell::RefCell,
    collections::HashMap,
    fmt::Debug,
    ops::{Deref, Index},
    rc::Rc,
};

pub mod base;
pub mod builder;
pub mod compare;
pub mod entity;
pub mod entity_ref;
pub mod entry;
pub mod props;
pub mod scopes;

pub use base::*;
pub use builder::*;
pub use entity::*;
pub use entity_ref::*;
pub use entry::*;
pub use props::*;
pub use scopes::*;

#[derive(Clone)]
pub struct EntityPtr(Rc<RefCell<Entity>>);

impl EntityPtr {
    pub fn new(e: Entity) -> Self {
        Self(Rc::new(RefCell::new(e)))
    }
}

impl Deref for EntityPtr {
    type Target = RefCell<Entity>;

    fn deref(&self) -> &Self::Target {
        self.0.as_ref()
    }
}

pub trait IntoEntry {
    fn to_entry(&self) -> Result<Entry, DomainError>;
}

impl IntoEntry for EntityRef {
    fn to_entry(&self) -> Result<Entry, DomainError> {
        use super::session::get_my_session;
        if !self.key().valid() {
            return Err(DomainError::InvalidKey);
        }
        get_my_session()?
            .entry(&LookupBy::Key(self.key()))?
            .ok_or(DomainError::DanglingEntity)
    }
}

use compare::{AnyChanges, CompareChanges, CompareError, Modified, Original};

pub fn any_entity_changes(
    l: AnyChanges<Option<Original>, EntityPtr>,
) -> Result<Option<Modified>, CompareError> {
    use compare::TreeDiff;

    let value_after = {
        let entity = l.after.borrow();

        serde_json::to_value(&*entity)?
    };

    let value_before: JsonValue = if let Some(original) = &l.before {
        match original {
            Original::String(s) => s.parse()?,
            Original::Json(v) => (*v).clone(),
        }
    } else {
        JsonValue::Null
    };

    let diff = TreeDiff {};

    diff.any_changes(AnyChanges {
        before: value_before,
        after: value_after,
    })
}
