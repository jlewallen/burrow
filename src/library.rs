use nom::{bytes::complete::take_while1, combinator::map, IResult};

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Item {
    Described(String),
}

pub fn word(i: &str) -> IResult<&str, &str> {
    take_while1(move |c| "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".contains(c))(i)
}

pub fn spaces(i: &str) -> IResult<&str, &str> {
    take_while1(move |c| " \t".contains(c))(i)
}

pub fn noun(i: &str) -> IResult<&str, Item> {
    map(word, |s: &str| Item::Described(s.to_owned()))(i)
}
