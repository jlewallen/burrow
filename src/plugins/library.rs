pub mod parser {
    pub use crate::kernel::*;
    pub use nom::{
        branch::alt,
        bytes::complete::{tag, take_while1},
        character::complete::digit1,
        combinator::map,
        combinator::{map_res, recognize},
        sequence::{pair, preceded, separated_pair},
        IResult,
    };
    pub use tracing::*;

    pub fn word(i: &str) -> IResult<&str, &str> {
        take_while1(move |c| "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".contains(c))(i)
    }

    pub fn spaces(i: &str) -> IResult<&str, &str> {
        take_while1(move |c| " \t".contains(c))(i)
    }

    pub fn noun(i: &str) -> IResult<&str, Item> {
        map(word, |s: &str| Item::Named(s.to_owned()))(i)
    }

    pub fn unsigned_number(i: &str) -> IResult<&str, u64> {
        map_res(recognize(digit1), str::parse)(i)
    }

    pub fn gid_reference(i: &str) -> IResult<&str, Item> {
        map(preceded(tag("#"), unsigned_number), |n| {
            Item::GID(EntityGID::new(n))
        })(i)
    }

    pub fn noun_or_specific(i: &str) -> IResult<&str, Item> {
        alt((noun, gid_reference))(i)
    }

    pub fn named_place(i: &str) -> IResult<&str, Item> {
        alt((
            gid_reference,
            map(word, |s: &str| Item::Route(s.to_owned())),
        ))(i)
    }
}

pub mod model {
    pub use crate::kernel::*;
    pub use anyhow::Result;
    pub use serde::{Deserialize, Serialize};
    pub use serde_json::{json, Value};
    pub use std::{collections::HashMap, ops::Deref};
    pub use tracing::*;
}

pub mod actions {
    pub use crate::kernel::*;
    pub use crate::plugins::log_test;
    pub use crate::plugins::tools;
    pub use anyhow::Result;
    pub use tracing::*;
}
