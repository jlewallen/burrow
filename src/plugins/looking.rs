pub mod model {
    use crate::plugins::library::model::*;
    use crate::{
        plugins::carrying::model::Containing,
        plugins::moving::model::{Movement, Occupyable},
    };

    #[derive(Debug, Serialize)]
    pub struct ObservedArea {}

    impl From<&Entity> for ObservedArea {
        fn from(_value: &Entity) -> Self {
            todo!()
        }
    }

    #[derive(Debug, Serialize)]
    pub struct ObservedPerson {}

    impl From<&Entity> for ObservedPerson {
        fn from(_value: &Entity) -> Self {
            todo!()
        }
    }

    #[derive(Debug, Serialize)]
    pub struct ObservedRoute {}

    impl From<&Entity> for ObservedRoute {
        fn from(_value: &Entity) -> Self {
            todo!()
        }
    }

    #[derive(Debug, Serialize)]
    pub struct ObservedEntity {
        pub key: EntityKey,
        pub name: Option<String>,
        pub desc: Option<String>,
    }

    impl From<Box<Entity>> for ObservedEntity {
        fn from(value: Box<Entity>) -> Self {
            Self {
                key: value.key.clone(),
                name: value.name(),
                desc: value.desc(),
            }
        }
    }

    impl From<&Entity> for ObservedEntity {
        fn from(value: &Entity) -> Self {
            Self {
                key: value.key.clone(),
                name: value.name(),
                desc: value.desc(),
            }
        }
    }

    // TODO This seems unnececssary, how can I help the compiler deduce the
    // proper chain of TryFrom/From to get here?
    impl TryFrom<&LazyLoadedEntity> for ObservedEntity {
        type Error = DomainError;

        fn try_from(value: &LazyLoadedEntity) -> Result<Self, Self::Error> {
            let entity = value.into_entity()?;
            let e = entity.borrow();

            Ok(Self {
                key: e.key.clone(),
                name: e.name(),
                desc: e.desc(),
            })
        }
    }

    #[derive(Debug, Serialize)]
    #[serde(rename_all = "camelCase")]
    pub struct AreaObservation {
        pub area: ObservedEntity,
        pub person: ObservedEntity,
        pub living: Vec<ObservedEntity>,
        pub items: Vec<ObservedEntity>,
        pub carrying: Vec<ObservedEntity>,
        pub routes: Vec<ObservedEntity>,
    }

    impl AreaObservation {
        pub fn new(user: &EntityPtr, area: &EntityPtr) -> Result<Self> {
            // I feel like there's a lot of unnecessary copying going on here.

            let mut living: Vec<ObservedEntity> = vec![];
            if let Ok(occupyable) = area.borrow().scope::<Occupyable>() {
                for entity in &occupyable.occupied {
                    living.push(entity.try_into()?);
                }
            }

            let mut items = vec![];
            if let Ok(containing) = area.borrow().scope::<Containing>() {
                for entity in &containing.holding {
                    items.push(entity.try_into()?);
                }
            }

            let mut carrying = vec![];
            if let Ok(containing) = user.borrow().scope::<Containing>() {
                for entity in &containing.holding {
                    carrying.push(entity.try_into()?);
                }
            }

            let mut routes = vec![];
            if let Ok(movement) = user.borrow().scope::<Movement>() {
                for route in &movement.routes {
                    routes.push((&route.area).try_into()?);
                }
            };

            Ok(AreaObservation {
                area: area.borrow().deref().into(),
                person: user.borrow().deref().into(),
                living,
                items,
                carrying,
                routes,
            })
        }
    }

    impl Reply for AreaObservation {
        fn to_markdown(&self) -> Result<Markdown> {
            let mut md = Markdown::new(Vec::new());
            md.write("")?;
            Ok(md)
        }
    }

    impl ToJson for AreaObservation {
        fn to_json(&self) -> Result<Value> {
            Ok(json!({ "areaObservation": serde_json::to_value(self)? }))
        }
    }

    #[derive(Debug, Serialize)]
    #[serde(rename_all = "camelCase")]
    pub struct InsideObservation {
        pub vessel: ObservedEntity,
        pub items: Vec<ObservedEntity>,
    }

    impl InsideObservation {
        pub fn new(_user: &EntityPtr, vessel: &EntityPtr) -> Result<Self> {
            let mut items = vec![];
            if let Ok(containing) = vessel.borrow().scope::<Containing>() {
                for entity in &containing.holding {
                    items.push(entity.try_into()?);
                }
            }

            Ok(InsideObservation {
                vessel: vessel.borrow().deref().into(),
                items,
            })
        }
    }

    impl Reply for InsideObservation {
        fn to_markdown(&self) -> Result<Markdown> {
            let mut md = Markdown::new(Vec::new());
            md.write("")?;
            Ok(md)
        }
    }

    impl ToJson for InsideObservation {
        fn to_json(&self) -> Result<Value> {
            Ok(json!({ "insideObservation": serde_json::to_value(self)? }))
        }
    }

    pub fn discover(_source: &Entity, _entity_keys: &mut [EntityKey]) -> Result<()> {
        Ok(())
    }
}

pub mod actions {
    use super::model::*;
    use super::parser::{parse, Sentence};
    use crate::plugins::library::actions::*;

    #[derive(Debug)]
    pub struct LookAction {}

    impl Action for LookAction {
        fn is_read_only() -> bool {
            true
        }

        fn perform(&self, (_world, user, area, _infra): ActionArgs) -> ReplyResult {
            info!("look!");

            Ok(Box::new(AreaObservation::new(&user, &area)?))
        }
    }

    #[derive(Debug)]
    pub struct LookInsideAction {
        item: Item,
    }

    impl Action for LookInsideAction {
        fn is_read_only() -> bool {
            true
        }

        fn perform(&self, args: ActionArgs) -> ReplyResult {
            info!("look inside!");

            let (_, user, _area, infra) = args.clone();

            match infra.find_item(args, &self.item)? {
                Some(target) => {
                    if tools::is_container(&target) {
                        Ok(Box::new(InsideObservation::new(&user, &target)?))
                    } else {
                        Ok(Box::new(SimpleReply::Impossible))
                    }
                }
                None => Ok(Box::new(SimpleReply::NotFound)),
            }
        }
    }

    pub fn evaluate(i: &str) -> EvaluationResult {
        Ok(parse(i).map(|(_, sentence)| evaluate_sentence(&sentence))?)
    }

    fn evaluate_sentence(s: &Sentence) -> Box<dyn Action> {
        match *s {
            Sentence::Look => Box::new(LookAction {}),
            Sentence::LookInside(_) => todo!(),
        }
    }

    #[cfg(test)]
    mod tests {
        use super::*;
        use crate::domain::{BuildActionArgs, QuickThing};

        #[test]
        fn it_looks_in_empty_area() -> Result<()> {
            let mut build = BuildActionArgs::new()?;
            let args: ActionArgs = build.plain().try_into()?;

            let action = LookAction {};
            let reply = action.perform(args.clone())?;
            let (_, _person, _area, _) = args.clone();

            insta::assert_json_snapshot!(reply.to_json()?);

            build.close()?;

            Ok(())
        }

        #[test]
        fn it_looks_in_area_with_items_on_ground() -> Result<()> {
            let mut build = BuildActionArgs::new()?;
            let args: ActionArgs = build
                .ground(vec![QuickThing::Object("Cool Rake".to_string())])
                .ground(vec![QuickThing::Object("Boring Shovel".to_string())])
                .try_into()?;

            let action = LookAction {};
            let reply = action.perform(args.clone())?;
            let (_, _person, _area, _) = args.clone();

            insta::assert_json_snapshot!(reply.to_json()?);

            build.close()?;

            Ok(())
        }

        #[test]
        fn it_looks_in_area_with_items_on_ground_and_a_route() -> Result<()> {
            let mut build = BuildActionArgs::new()?;
            let destination = build.make(QuickThing::Place("Place".to_string()))?;
            let args: ActionArgs = build
                .ground(vec![QuickThing::Object("Cool Rake".to_string())])
                .ground(vec![QuickThing::Object("Boring Shovel".to_string())])
                .route("East Exit", QuickThing::Actual(destination.clone()))
                .try_into()?;

            let action = LookAction {};
            let reply = action.perform(args.clone())?;
            let (_, _person, _area, _) = args.clone();

            insta::assert_json_snapshot!(reply.to_json()?);

            build.close()?;

            Ok(())
        }

        #[test]
        fn it_looks_in_area_with_items_on_ground_and_holding_items() -> Result<()> {
            let mut build = BuildActionArgs::new()?;
            let destination = build.make(QuickThing::Place("Place".to_string()))?;
            let args: ActionArgs = build
                .ground(vec![QuickThing::Object("Boring Shovel".to_string())])
                .hands(vec![QuickThing::Object("Cool Rake".to_string())])
                .route("East Exit", QuickThing::Actual(destination.clone()))
                .try_into()?;

            let action = LookAction {};
            let reply = action.perform(args.clone())?;
            let (_, _person, _area, _) = args.clone();

            insta::assert_json_snapshot!(reply.to_json()?);

            build.close()?;

            Ok(())
        }

        #[test]
        fn it_fails_to_look_inside_non_containers() -> Result<()> {
            let mut build = BuildActionArgs::new()?;
            let args: ActionArgs = build
                .hands(vec![QuickThing::Object("Not A Box".to_string())])
                .try_into()?;

            let action = LookInsideAction {
                item: Item::Named("box".to_owned()),
            };
            let reply = action.perform(args.clone())?;
            let (_, _person, _area, _) = args.clone();

            insta::assert_json_snapshot!(reply.to_json()?);

            build.close()?;

            Ok(())
        }

        #[test]
        fn it_looks_inside_containers() -> Result<()> {
            let mut build = BuildActionArgs::new()?;
            let vessel = build
                .build()?
                .named("Vessel")?
                .holding(&vec![build.make(QuickThing::Object("Key".to_string()))?])?
                .into_entity()?;
            let args: ActionArgs = build.hands(vec![QuickThing::Actual(vessel)]).try_into()?;

            let action = LookInsideAction {
                item: Item::Named("vessel".to_owned()),
            };
            let reply = action.perform(args.clone())?;
            let (_, _person, _area, _) = args.clone();

            insta::assert_json_snapshot!(reply.to_json()?);

            build.close()?;

            Ok(())
        }
    }
}

pub mod parser {
    use crate::plugins::library::parser::*;

    // TODO Underneath, Above, Behond, etc... 'Physically Relative'
    #[derive(Debug, Clone, Eq, PartialEq)]
    pub enum Sentence {
        Look,
        LookInside(Item),
    }

    pub fn parse(i: &str) -> IResult<&str, Sentence> {
        look(i)
    }

    fn look(i: &str) -> IResult<&str, Sentence> {
        let inside = map(
            separated_pair(
                separated_pair(tag("look"), spaces, tag("inside")),
                spaces,
                noun,
            ),
            |(_, nearby)| Sentence::LookInside(nearby),
        );

        let area = map(tag("look"), |_| Sentence::Look);

        alt((inside, area))(i)
    }

    #[cfg(test)]
    mod tests {
        use super::*;

        #[test]
        fn it_parses_look_correctly() {
            let (remaining, actual) = parse("look").unwrap();
            assert_eq!(remaining, "");
            assert_eq!(actual, Sentence::Look)
        }

        #[test]
        fn it_parses_look_inside_correctly() {
            let (remaining, actual) = parse("look inside box").unwrap();
            assert_eq!(remaining, "");
            assert_eq!(actual, Sentence::LookInside(Item::Named("box".to_owned())))
        }

        #[test]
        fn it_errors_on_unknown_text() {
            let output = parse("hello");
            assert!(output.is_err());
        }
    }
}
