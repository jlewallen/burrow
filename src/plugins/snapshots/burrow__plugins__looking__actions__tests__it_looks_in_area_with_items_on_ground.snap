---
source: src/plugins/looking.rs
expression: reply.to_json()?
---
{
  "areaObservation": {
    "area": {
      "desc": "Not described.",
      "key": "E-2",
      "name": "Welcome Area"
    },
    "carrying": [],
    "items": [
      {
        "desc": "Not described.",
        "key": "E-3",
        "name": "Cool Rake"
      },
      {
        "desc": "Not described.",
        "key": "E-4",
        "name": "Boring Shovel"
      }
    ],
    "living": [
      {
        "desc": "Not described.",
        "key": "E-1",
        "name": "Living"
      }
    ],
    "person": {
      "desc": "Not described.",
      "key": "E-1",
      "name": "Living"
    },
    "routes": []
  }
}
